<?xml version="1.0"?>

<gazebo:world 
  xmlns:xi="http://www.w3.org/2001/XInclude"
  xmlns:gazebo="http://playerstage.sourceforge.net/gazebo/xmlschema/#gz" 
  xmlns:model="http://playerstage.sourceforge.net/gazebo/xmlschema/#model" 
  xmlns:sensor="http://playerstage.sourceforge.net/gazebo/xmlschema/#sensor" 
  xmlns:window="http://playerstage.sourceforge.net/gazebo/xmlschema/#window" 
  xmlns:param="http://playerstage.sourceforge.net/gazebo/xmlschema/#param" 
  xmlns:body="http://playerstage.sourceforge.net/gazebo/xmlschema/#body" 
  xmlns:geo="http://willowgarage.com/xmlschema/#geo" 
  xmlns:geom="http://playerstage.sourceforge.net/gazebo/xmlschema/#geom" 
  xmlns:joint="http://playerstage.sourceforge.net/gazebo/xmlschema/#joint" 
  xmlns:interface="http://playerstage.sourceforge.net/gazebo/xmlschema/#interface" 
  xmlns:ui="http://playerstage.sourceforge.net/gazebo/xmlschema/#ui"
  xmlns:rendering="http://playerstage.sourceforge.net/gazebo/xmlschema/#rendering" 
  xmlns:renderable="http://playerstage.sourceforge.net/gazebo/xmlschema/#renderable" 
  xmlns:controller="http://playerstage.sourceforge.net/gazebo/xmlschema/#controller"
  xmlns:physics="http://playerstage.sourceforge.net/gazebo/xmlschema/#physics" >

  <thread>4</thread>
  <verbosity>5</verbosity>

  <!-- cfm is 1e-5 for single precision -->
  <!-- erp is typically .1-.8 -->
  <!-- here's the global contact cfm/erp -->
  <physics:ode>
    <stepTime>0.001</stepTime>
    <gravity>0 0 -9.8</gravity>
    <cfm>0.0000000001</cfm>
    <erp>0.2</erp>
    <quickStep>true</quickStep>
    <quickStepIters>10</quickStepIters>
    <quickStepW>1.3</quickStepW>
    <contactMaxCorrectingVel>100.0</contactMaxCorrectingVel>
    <contactSurfaceLayer>0.001</contactSurfaceLayer>
  </physics:ode>

  <geo:origin>
    <lat>37.4270909558</lat><lon>-122.077919338</lon>
  </geo:origin>

  <rendering:gui>
    <type>fltk</type>
    <size>480 320</size>
    <pos>0 0</pos>
    <frames>
      <row height="100%">
        <camera width="100%">
          <xyz>0 0 2.0</xyz>
          <rpy>-5.0 6.75 1.8</rpy>
        </camera>
      </row>
    </frames>
  </rendering:gui>

  <rendering:ogre>
    <ambient>0.8 0.8 0.8 0.8</ambient>
    <sky>
      <material>Gazebo/CloudySky</material>
    </sky>
    <grid>false</grid>
    <maxUpdateRate>10.</maxUpdateRate>
    <shadowTechnique>stencilAdditive</shadowTechnique>
  </rendering:ogre>

  <!-- clock -->
  <model:physical name="clock">
    <xyz>0 0 0</xyz>
    <rpy>0 0 0</rpy>
    <static>true</static>
    <body:box name="clock_body">
      <geom:box name="clock_geom">
        <mesh>default</mesh>
        <size>0 0 0</size>
        <visual>
          <size>0 0 0</size>
          <material>Gazebo/White</material>
          <mesh>unit_box</mesh>
        </visual>
      </geom:box>
    </body:box>

    <!-- broadcast ros /time with simTime -->
    <controller:gazebo_ros_time name="gazebo_ros_time" plugin="libgazebo_ros_time.so">
      <alwaysOn>true</alwaysOn>
      <updateRate>1000.0</updateRate>
      <interface:audio name="dummy_iface1"/>
    </controller:gazebo_ros_time>
  
    <controller:gazebo_ros_factory name="gazebo_ros_factory" plugin="libgazebo_ros_factory.so">
      <alwaysOn>true</alwaysOn>
      <updateRate>20.0</updateRate>
      <robotNamespace>/</robotNamespace>
      <spawnModelServiceName>spawn_model</spawnModelServiceName>
      <deleteModelServiceName>bad_delete_model</deleteModelServiceName>
      <interface:audio name="dummy_iface2"/>
    </controller:gazebo_ros_factory>

    <controller:delete_object_plugin name="delete_object_plugin" plugin="libdelete_object_plugin.so">
        <alwaysOn>true</alwaysOn>
        <updateRate>20.0</updateRate>
        <interface:audio name="dummy_iface3"/>
    </controller:delete_object_plugin>

    <controller:add_object_plugin name="add_object_plugin" plugin="libadd_object_plugin.so">
        <alwaysOn>true</alwaysOn>
        <updateRate>20.0</updateRate>
        <interface:audio name="dummy_iface4"/>
    </controller:add_object_plugin>

    <controller:world_state_publisher name="world_state_publisher" plugin="libworld_state_publisher.so">
        <alwaysOn>true</alwaysOn>
        <topicName>gazebo_world_state</topicName>
        <updateRate>10.0</updateRate>
        <interface:audio name="world_state_dummy_iface" />
    </controller:world_state_publisher>
  </model:physical>

  <!-- ground plane -->
    <model:physical name="gplane">
        <xyz>0 0 0</xyz>	
        <rpy>0 0 0</rpy>
        <static>true</static>

        <body:plane name="plane">
          <geom:plane name="plane">
            <laserRetro>2000.0</laserRetro>
            <mu1>50.0</mu1>
            <mu2>50.0</mu2>
            <kp>1000000000.0</kp>
            <kd>1.0</kd>
            <normal>0 0 1</normal>
            <size>51.3 51.3</size>
            <segments>10 10</segments>
            <uvTile>100 100</uvTile>
            <material>Gazebo/GrayGrid</material>
          </geom:plane>
        </body:plane>
    </model:physical>

    <!-- White Point light -->
    <model:renderable name="point_white">
        <xyz>0.0 0.0 3</xyz>
        <enableGravity>false</enableGravity>
        <light>
            <type>point</type>
            <diffuseColor>0.5 0.5 0.5</diffuseColor>
            <specularColor>.1 .1 .1</specularColor>
            <attenuation>0.2 0.1 0</attenuation>
            <range>10</range>
        </light>
    </model:renderable>

</gazebo:world>
